### ISSUE TITLE
Resend failed messages in open campaign

### ISSUE THREAD

**Issue by wjwoodson** [gophish/gophish#79]
Mail servers can reject messages (temporary or permanent SMTP message) to enforce rate limits, etc. From campaign detail screen, allow admin user to attempt re-sending messages in a send failed status.

**Comment by jordan-wright** [comment]
I'd like to handle this the same way [sidekiq](https://github.com/mperham/sidekiq/wiki/Error-Handling#automatic-job-retry) does, with an exponential back off. After *x* failures, just set the status to send failure or something.

**Comment by jordan-wright** [comment]
I'd also like to be able to manually re-send an email from the campaign results screen.

**Comment by wjwoodson** [comment]
Should be able to work on this with stored sending added #169

**Comment by jordan-wright** [comment]
Whoops, I may have responded to wrong issue. This sounds good!

**Comment by jordan-wright** [comment]
Yep! I imagine a button on the timeline that allows you to resend to a specific user. It can be a special API endpoint, like /campaign/:id/resend that takes in a campaign result object.

**Comment by jordan-wright** [comment]
If you're going to start working on this, go ahead and make a new branch since I'm considering master pretty well locked for 0.1.2. I'm just shoring up docs now.

**Comment by StrangeUSB** [comment]
Wait are you wanting to resend the whole campaign or just the failed
messages?

**Comment by jordan-wright** [comment]
I think resending the whole campaign should just be the "copy campaign" function, since it would likely be initiated because the user put in a wrong config option.

**Comment by StrangeUSB** [comment]
Ok, yeah I was thinking that it would be nice to have a couple options.

Resend a specific message
Resend all failed/errored messages
Resend full campaign

**Comment by wjwoodson** [comment]
I would think we'd handle resend per failed message in the api. Js may
iterate over failed messages at campaign level, but resending an entire
campaign should definitely be done via campaign copy.

Also may want to do the exponential backoff per message, need to determine
best way to handle that without impacting application if bad sender is
specified or rate limiting (retry) happens at configured SMTP

**Comment by jordan-wright** [comment]
> Also may want to do the exponential backoff per message, need to determine
best way to handle that without impacting application if bad sender is
specified or rate limiting (retry) happens at configured SMTP

Agreed completely. That would be another issue (I think it already is?). That's going to be a bit tricky.

Iterating over the single API endpoint sounds like a great idea to me!

**Comment by cclements** [comment]
Adding in that it would be great to be able to do a campaign wide re-send based on status (e.g., resend to all with status of "unopened", or all with status of "unopened" and "opened" but not "clicked link/submitted data")

**Comment by bfreitasdtiq** [comment]
Any News on this implementation? OR a way to schedule the time between sent emails

**Comment by acklenx** [comment]
Just ran into this today.  Sent a campaign out to 600 recipients and just over half had an error:

   Max connection attempts exceeded - dial tcp 44.191.248.250:25: i/o timeout

That was using MailGun (fake IP listed) with SMTP.  I would love to be able to resend to all the errors without having to create a new campaign.  Which sadly involves collecting all the errored email address first (they're all spread out).

And while I'm here, let me also up vote the related feature to resend to onclicked/unopened addersses too.  That would be super helpful for every day simulations when I don't have errors.

**Comment by acklenx** [comment]
I should have mentioned that I scheduled that batch of 660 messages to be send over the course of 30 minutes.  There were batches of errors from 20-60 or so at a time

I waited (collected new list of failed addresses) about an hour or so and resent to all 300 ish failed addresses, this time with no delay.  Less than 2 minutes later they all finished with 0 errors!

**Comment by MKprogrammer** [comment]
Hey ! 
First of All i want to say thanks for this grate project!
Now, There is any update for the resend button in the campaign page?
I read in some threads that this could be reached via the API but i couldn't find that there..
@jordan-wright Thanks again for your great work!

**Comment by infosecconsultant** [comment]
Hey all, 

Also looking for an update to this feature as I've kicked off a campaign where it's hit mail-server rate limiting about halfway through (130 emails of ~260).
The software has waited a minute and then tried to re-send but this has resulted in the error message changing from

"Error 421 4.7.0 Try again later, closing connection. (MAIL)  - gsmtp"
to 
"Error Max connection attempts exceeded - 454 4.7.0 Too many login attempts, please try again later. - gsmtp"

(Am obviously using GSuite here to mail via).

Rather than extracting the list of users that have failed to send, it'd be nice even  if I could click through them and manually hit 'retry/resend'.

Thanks!

**Comment by ironc0in** [comment]
i guess a lot ppl are facing this issue. I'm in the same boat. Even though I used the feature to spread email sending across 3-4 days, which essentially was sending 1 email per 30 mins, when a few of them failed, it backlogged my mail server and the mail server started sending them all out with 30 sec intervals, which caused spam filters to hold it longer!

**Comment by twosevenzero** [comment]
Any updates on this one or anything the community can do to help?

**Comment by vincentcox** [comment]
Any updates on this?

**Comment by emixam3** [comment]
Hi,
Any update on this feature?

**Comment by glennzw** [comment]
No updates as of yet - but open to PRs or a discussion to get the dev ball rolling.

**Comment by murilo-melo-caramelo** [comment]
Any updates?

**Comment by edgardok** [comment]
hi! any updates?

**Comment by pratyakshsl7** [comment]
Any updates??

**Comment by ocervell** [comment]
any updates on this ? it would be great to be able to individually resend messages as well

**Comment by Thrash0ld1** [comment]
OMG so many years for simple and so useful feature. why this is not implemented...
